main.go

```diff
@@ -114,7 +114,7 @@ func main() {
 	case "aws":
 		p, err = provider.NewAWSProvider(cfg.DomainFilter, cfg.DryRun)
 	case "inmemory":
-		p, err = provider.NewInMemoryProviderWithDomainAndLogging("example.com"), nil
+		p, err = provider.NewInMemoryProvider(provider.InMemoryWithDomain(cfg.DomainFilter), provider.InMemoryWithLogging()), nil
```

Comment:
I think extending via `options` is more typical. Didn't quite get how `OnError()` would work ? 

The last one I think is unusual, calling `Do()` on initializer